### PR TITLE
fix subgroup cost model

### PR DESCRIPTION
**Pull request by adstraw**
Roughly a 10% speedup on ResNet50.  Kernel times dropped from ~63ms to ~57ms on my machine.

### CODE REVIEW

File: pmlc/dialect/pxa/transforms/subgroups.cc [hunk 1]
```diff
@@ -159,58 +175,153 @@ struct SubgroupCostModel {
...
+      for (auto kvp : positionalDimStrides.strides) {
+        auto index = getIndexByArgument(kvp.first);
+        auto dimStride = kvp.second;
+
+        // dim strides should be a superset of flat strides
+        assert(flatStrides.strides.count(kvp.first));
+        auto flatStride = flatStrides.strides[kvp.first];
+
+        // set tensor stride the first time through the loop
+        if (!hasTensorStride) {
```
Comment (jbruestle) [inline pmlc/dialect/pxa/transforms/subgroups.cc:250]: We can get tensor stride at precompute time (when we get dimStrides + flatStrides) via a call to getStridesAndOffset (see Stride.cc flatten for an example).

**Review by jbruestle — CHANGES_REQUESTED** [review]
Looks good!  Some minor change requests.

File: pmlc/dialect/pxa/transforms/subgroups.cc [hunk 2]
```diff
@@ -159,58 +175,153 @@ struct SubgroupCostModel {
...
+    double cacheMiss;
+    SmallVector<int64_t, 4> tensorStrides;
+    SmallVector<int64_t, 4> tensorDimensions;
   };
 
-  // Compute the memory info for a single load/store given the current plan
-  MemInfo computeMemoryInfo(StrideInfo si) {
-    MemInfo out = {0, 1};
-    for (unsigned i = 0; i < ranges.size(); i++) {
+  size_t getIndexByArgument(mlir::BlockArgument arg) {
```
[the remark this answers is not in the document]
Comment (adstraw) [reply pmlc/dialect/pxa/transforms/subgroups.cc:214]: Done.

File: pmlc/dialect/pxa/transforms/subgroups.cc [hunk 3]
```diff
@@ -159,58 +175,153 @@ struct SubgroupCostModel {
     return;
   }
 
+  double memoryIO(double cacheElems, SmallVector<int64_t, 4> tensorDimensions,
```
[the remark this answers is not in the document]
Comment (adstraw) [reply pmlc/dialect/pxa/transforms/subgroups.cc]: Done, moved to strides.cc

File: pmlc/dialect/pxa/transforms/subgroups.cc (hunk 1 above)
Comment (adstraw) [reply pmlc/dialect/pxa/transforms/subgroups.cc]: Done.

**Review by jbruestle — APPROVED** [review]

File: pmlc/dialect/pxa/transforms/subgroups.cc [hunk 4]
```diff
@@ -120,23 +120,30 @@ struct SubgroupCostModel {
...
+    int64_t offset;
+    SmallVector<int64_t, 4> tensorStrides;
+    if (failed(getStridesAndOffset(memreftype, tensorStrides, offset))) {
+      IVLOG(3, "Cannot compute tensor strides");
+      return false;
+    }
+    ioTensorStrides.push_back(tensorStrides);
+    IVLOG(3, "  dimensional strides = " << tensorStrides);
+
+    auto elementtype = memreftype.getElementType();
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/subgroups.cc:144]: Rename to `elementType`

**Pull request merged**